### PR TITLE
issue 165: more ods types detected

### PR DESCRIPTION
**Pull request by chfw**
Current ODSRowSet only detects single line text if it is lucky because the cell type detection

    cell_type = elem.attrib.get('urn:oasis:names:tc:opendocument:xmlns:office:1.0:value-type')

should have been

    cell_type = elem.attrib.get('{urn:oasis:names:tc:opendocument:xmlns:office:1.0}value-type')

And text cell could be of multiple lines. For example, the cell could have multiple tags instead of one. Current implementation read the first tag and stopped.

Without guessing, it is possible to detect from ods specification without guessing:

- currency
- percentage
- time and time delta
- boolean

And 'number-columns-repeated', which allows the value of a column to be repeated, could be detected as well.

### CODE REVIEW

File: test/test_read.py [hunk 1]
```diff
@@ -308,9 +311,9 @@ def test_read_large_ods(self):
         assert_equal(6, len(table_set.tables))
         row_set = table_set.tables[0]
         row = next(row_set.raw())
-        assert len(row) == 5, len(row)
```
Comment (chfw) [inline test/test_read.py]: with number-columns-repeated, this becomes 16384.

File: messytables/ods.py [hunk 2]
```diff
@@ -129,19 +154,61 @@ def raw(self, sample=False):
             block = self.namespace_tags[0] + row + self.namespace_tags[1]
             partial = io.BytesIO(block)
 
-            for action, elem in etree.iterparse(partial, ('end',)):
-                if elem.tag == '{urn:oasis:names:tc:opendocument:xmlns:table:1.0}table-cell':
-                    cell_type = elem.attrib.get('urn:oasis:names:tc:opendocument:xmlns:office:1.0:value-type')
```
Comment (chfw) [inline messytables/ods.py]: cell_type is always None unfortunately.

File: messytables/ods.py [hunk 3]
```diff
@@ -102,13 +127,13 @@ def __init__(self, sheet, window=None, namespace_tags=None):
         else:
             namespaces = {
                 "dc": u"http://purl.org/dc/elements/1.1/",
-                "draw": u"urn:oasis:names:tc:opendocument:xmlns:drawing:1.0",
-                "number": u"urn:oasis:names:tc:opendocument:xmlns:datastyle:1.0",
-                "office": u"urn:oasis:names:tc:opendocument:xmlns:office:1.0",
-                "svg": u"urn:oasis:names:tc:opendocument:xmlns:svg-compatible:1.0",
-                "table": u"urn:oasis:names:tc:opendocument:xmlns:table:1.0",
-                "text": u"urn:oasis:names:tc:opendocument:xmlns:text:1.0",
-                "calcext": u"urn:org:documentfoundation:names:experimental:calc:xmlns:calcext:1.0",
```
Comment (chfw) [inline messytables/ods.py]: exceeding 80 characters.

**Comment by rufuspollock** [comment]
@chfw these look like a bunch of useful fixes. Could you rebase against master quickly and we can see if we can get these in.

File: messytables/types.py [hunk 4]
```diff
@@ -134,6 +155,25 @@ def cast(self, value):
...
+            return value
+        if value in ('', None):
+            return None
+        hour = int(value[2:4])
+        minute = int(value[5:7])
+        second = int(value[8:10])
+        if hour < 24:
+            return datetime.time(hour, minute, second)
+        else:
+            return datetime.timedelta(hours=hour,
```
Comment (chfw) [inline messytables/types.py]: Here is relevant info regarding TimeDelta: https://github.com/pyexcel/pyexcel-ods3/commit/083b0aa4701bd0997e9fd5638c53758935ca41fb. Basically, ods support time delta where the hour value is greater than or equal to 24.

**Comment by rufuspollock** [comment]
Also assume you are happy to license your contributions here under relevant license :-)

**Comment by chfw** [comment]
MIT is fine with me.

**Comment by fxsman** [comment]
We need this datetime  type for filtering in tables, can anyone confirm and push this, please?

**Pull request merged**